File: room-demo-rxandroid-livedata/src/main/java/com/example/pavneet_singh/room_demo_rxandroid/ui/AddNoteActivity.java
package com.example.pavneet_singh.room_demo_rxandroid.ui;

import android.os.Bundle;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.pavneet_singh.room_demo_rxandroid.R;
import com.example.pavneet_singh.room_demo_rxandroid.notedb.NoteDatabase;
import com.example.pavneet_singh.room_demo_rxandroid.notedb.model.Note;
import com.google.android.material.textfield.TextInputEditText;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

public class AddNoteActivity extends AppCompatActivity {

    private TextInputEditText et_title, et_content;
    private NoteDatabase noteDatabase;
    private Note note;
    private boolean update;
    private CompositeDisposable compositeDisposable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_note);
        initUI();
    }

    private void initUI() {
        compositeDisposable = new CompositeDisposable();
        et_title = findViewById(R.id.et_title);
        et_content = findViewById(R.id.et_content);
        noteDatabase = NoteDatabase.getInstance(AddNoteActivity.this);
        Button button = findViewById(R.id.but_save);
        if ((note = (Note) getIntent().getSerializableExtra("note")) != null) {
            getSupportActionBar().setTitle("Update Note");
            update = true;
            button.setText("Update");
            et_title.setText(note.getTitle());
            et_content.setText(note.getContent());
        }
        button.setOnClickListener(view -> {
            if (update) {
                note.setContent(et_content.getText().toString());
                note.setTitle(et_title.getText().toString());
                compositeDisposable.add(noteDatabase.getNoteDao().updateNote(note)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(this::finish, e -> {
                            Toast.makeText(this, "Update failed due to " + e.getMessage(), Toast.LENGTH_SHORT).show();
                            e.printStackTrace();
                        }));
            } else {
                note = new Note(et_content.getText().toString(), et_title.getText().toString());
                compositeDisposable.add(noteDatabase.getNoteDao().insertNote(note)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        // Can replace rowId -> finish() with this::finish by changing Single to Completable in DAO for insert
                        .subscribe(rowId -> finish(), Throwable::printStackTrace));
            }
        });
    }

    @Override
    protected void onDestroy() {
        compositeDisposable.dispose();
        super.onDestroy();
    }
}

File: room-demo-kotin-mvvm-dagger/src/main/java/com/example/pavneet_singh/room_demo_kotin_mvvm_dagger/ui/AddNoteActivity.kt
package com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.ui

import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.R
import com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.notedb.NoteDataBase
import com.example.pavneet_singh.room_demo_kotin_mvvm_dagger.notedb.model.Note
import com.google.android.material.textfield.TextInputEditText
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

/**
 * Created by Pavneet_Singh on 2020-01-25.
 */

class AddNoteActivity : AppCompatActivity() {
    private lateinit var etTitle: TextInputEditText
    private lateinit var etContent: TextInputEditText
    private lateinit var noteDatabase: NoteDataBase
    private lateinit var note: Note
    private var update = false
    private lateinit var compositeDisposable: CompositeDisposable

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_note)
        initUI()
    }

    private fun initUI() {
        compositeDisposable = CompositeDisposable()
        etTitle = findViewById(R.id.et_title)
        etContent = findViewById(R.id.et_content)
        noteDatabase = NoteDataBase.getInstance(this@AddNoteActivity)
        val button = findViewById<Button>(R.id.but_save)

        intent.getSerializableExtra("note")?.let {
            (it as Note).apply {
                note = this
                supportActionBar!!.title = "Update Note"
                update = true
                button.text = "Update"
                etTitle.setText(title)
                etContent.setText(content)
            }
        }
        button.setOnClickListener {
            if (update) {
                note.content = etContent.text.toString()
                note.title = etTitle.text.toString()
                compositeDisposable.add(noteDatabase.getNoteDao().updateNote(note)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                        { finish() },
                        { e ->
                            Toast.makeText(
                                this,
                                "Update failed due to " + e.message,
                                Toast.LENGTH_SHORT
                            ).show()
                            e.printStackTrace()
                        }
                    ))
            } else {
                note = Note(
                    content = etContent.text.toString(), // using placeholder name to avoid passing long id, as using default as 0
                    title = etTitle.text.toString()
                )
                compositeDisposable.add(noteDatabase.getNoteDao().insertNote(note)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                        { rowId ->
                            note.note_id = rowId // update rowId in database
                            finish()
                        },
                        { obj: Throwable -> obj.printStackTrace() }
                    ))
            }
        }
    }

    override fun onDestroy() {
        compositeDisposable.dispose()
        super.onDestroy()
    }
}